store/tikv/coprocessor.go

```diff
@@ -107,35 +109,161 @@ const (
-	ranges := make([]*coprocessor.KeyRange, 0, len(t.ranges))
-	for _, r := range t.ranges {
+// copRanges is like []kv.KeyRange, but may has extra elements at head/tail.
+// It's for avoiding alloc big slice during build copTask.
+type copRanges struct {
+	first *kv.KeyRange
+	mid   []kv.KeyRange
+	last  *kv.KeyRange
+}
+
+func (r *copRanges) String() string {
+	var s string
+	r.do(func(ran *kv.KeyRange) {
+		s += fmt.Sprintf("[%q, %q]", ran.StartKey, ran.EndKey)
+	})
+	return s
+}
+
+func (r *copRanges) len() int {
+	var l int
+	if r.first != nil {
+		l += 1
+	}
+	l += len(r.mid)
+	if r.last != nil {
+		l += 1
+	}
+	return l
+}
+
+func (r *copRanges) at(i int) kv.KeyRange {
+	if r.first != nil {
+		if i == 0 {
+			return *r.first
+		}
+		i--
+	}
+	if i < len(r.mid) {
+		return r.mid[i]
+	}
+	return *r.last
+}
+
+func (r *copRanges) slice(from, to int) *copRanges {
+	var ran copRanges
+	if r.first != nil {
+		if from == 0 && to > 0 {
+			ran.first = r.first
+		}
+		if from > 0 {
+			from--
+		}
+		if to > 0 {
+			to--
+		}
+	}
+	if to <= len(r.mid) {
+		ran.mid = r.mid[from:to]
+	} else {
+		ran.mid, ran.last = r.mid[from:], r.last
+	}
+	return &ran
+}
+
+func (r *copRanges) do(f func(ran *kv.KeyRange)) {
+	if r.first != nil {
+		f(r.first)
+	}
+	for _, ran := range r.mid {
+		f(&ran)
+	}
+	if r.last != nil {
+		f(r.last)
+	}
+}
+
+func (r *copRanges) toPBRanges() []*coprocessor.KeyRange {
+	ranges := make([]*coprocessor.KeyRange, 0, r.len())
+	r.do(func(ran *kv.KeyRange) {
 		ranges = append(ranges, &coprocessor.KeyRange{
-			Start: r.StartKey,
-			End:   r.EndKey,
+			Start: ran.StartKey,
+			End:   ran.EndKey,
 		})
-	}
+	})
 	return ranges
 }
 
-func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges []kv.KeyRange, desc bool) ([]*copTask, error) {
+func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges *copRanges, desc bool) ([]*copTask, error) {
+	start := time.Now()
+	rangesLen := ranges.len()
+
 	var tasks []*copTask
-	for _, r := range ranges {
-		var err error
-		if tasks, err = appendTask(tasks, bo, cache, r); err != nil {
+	appendTask := func(region *Region, ranges *copRanges) {
+		tasks = append(tasks, &copTask{
+			idx:      len(tasks),
+			region:   region,
+			status:   taskNew,
+			ranges:   ranges,
+			respChan: make(chan *coprocessor.Response, 1),
+		})
+	}
+
+	for ranges.len() > 0 {
+		region, err := cache.GetRegion(bo, ranges.at(0).StartKey)
+		if err != nil {
 			return nil, errors.Trace(err)
 		}
+
+		// Iterate to the first range that is not complete in the region.
+		var i int
+		for ; i < ranges.len(); i++ {
+			r := ranges.at(i)
+			if !(region.Contains(r.EndKey) || bytes.Equal(region.EndKey(), r.EndKey)) {
+				break
+			}
+		}
+		// All rest ranges belong to the same region.
+		if i == ranges.len() {
+			appendTask(region, ranges)
+			break
+		}
+
+		r := ranges.at(i)
+		if region.Contains(r.StartKey) {
+			// Part of r is not in the region. We need to split it.
```

Comment:
if bytes.Equal(region.EndKey(), r.EndKey) && region.Contains(r.StartKey), the whole r is in region.